### PR TITLE
[function] compatible with mysql when aggregate function work on empty data

### LINKED ISSUE #771: Return default value when aggregating empty data
(you don't have to strictly follow this form)

**Describe the bug**

```
mysql> select sum(number) from numbers(100) where number > 1000;
Query OK, 0 rows affected (0.01 sec)

mysql> select sumIf(number, number > 1000) from numbers(100);
Query OK, 0 rows affected (0.00 sec)

```

Expect to return null, https://www.db-fiddle.com/f/qDnLWye3pM1thKfx4d8Z2M/0

### PR DESCRIPTION
**Pull request by zhaox1n**
I hereby agree to the terms of the CLA available at: https://datafuse.rs/policies/cla/

## Summary

Summary about this PR

## Changelog

- Bug Fix
- Improvement

## Related Issues

Fixes #771

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: common/aggregate_functions/src/aggregate_arg_max.rs [hunk 1]
```diff
@@ -59,6 +59,9 @@ impl AggregateFunction for AggregateArgMaxFunction {
     }
 
     fn accumulate(&mut self, columns: &[DataColumnarValue], _input_rows: usize) -> Result<()> {
+        if _input_rows == 0 {
```
Comment (zhang2014) [inline common/aggregate_functions/src/aggregate_arg_max.rs:62]: _input_rows -> input_rows maybe is better.

File: common/aggregate_functions/src/aggregate_function_test.rs [hunk 2]
```diff
@@ -159,3 +159,143 @@ fn test_aggregate_function() -> Result<()> {
...
+            error: "",
+        },
+        Test {
+            name: "uniq-passed",
+            eval_nums: 1,
+            args: vec![args[0].clone()],
+            display: "uniq",
+            func_name: "uniq",
+            columns: vec![columns[0].clone()],
+            expect: DataValue::UInt64(Some(0)),
```
Comment (zhang2014) [inline common/aggregate_functions/src/aggregate_function_test.rs:264]: maybe uniq function should return empty rows?

File: common/aggregate_functions/src/aggregate_arg_max.rs (hunk 1 above)
Comment (zhaox1n) [reply common/aggregate_functions/src/aggregate_arg_max.rs]: Done

File: tests/suites/0_stateless/03_0002_select_aggregator_with_filter.sql [hunk 3]
```diff
@@ -1,2 +1,9 @@
 SELECT max(number) FROM numbers_mt (10) where number > 99999999998;
 SELECT max(number) FROM numbers_mt (10) where number > 2;
+
+select min(number) from numbers_mt (10) where 1 = 2;
```
Comment (zhang2014) [inline tests/suites/0_stateless/03_0002_select_aggregator_with_filter.sql:4]: What happens to the following queries?
- `SELECT min(number) FROM numbers_mt(10) WHERE 1 = 2 GROUP BY number % 3`
- `SELECT number % 3, min(number) FROM numbers_mt(10) WHERE 1 = 2 GROUP BY number % 3`

Comment (zhang2014) [reply tests/suites/0_stateless/03_0002_select_aggregator_with_filter.sql:4]: MySQL 5.7 behavior:
https://www.db-fiddle.com/f/qDnLWye3pM1thKfx4d8Z2M/1

Comment (zhaox1n) [reply tests/suites/0_stateless/03_0002_select_aggregator_with_filter.sql:4]: No row returns. I will add related test later.

File: common/aggregate_functions/src/aggregate_function_test.rs (hunk 2 above)
Comment (zhaox1n) [reply common/aggregate_functions/src/aggregate_function_test.rs:264]: The uniq function seems to behave the same as count distinct, so I return 0

File: tests/suites/0_stateless/02_00007_function_aggreator_on_empty.result [hunk 4]
```diff
@@ -0,0 +1,16 @@
+
```
Comment (sundy-li) [inline tests/suites/0_stateless/02_00007_function_aggreator_on_empty.result:1]:
```
/// Get the value at the given row in an array as a String.
///
/// Note this function is quite inefficient and is unlikely to be
/// suitable for converting large arrays or record batches.
pub fn array_value_to_string(column: &array::ArrayRef, row: usize) -> Result<String> {
    if column.is_null(row) {
        return Ok("".to_string());
    }
```

Seems the behavior in arrow `array_value_to_string` is wrong.

Comment (zhaox1n) [reply tests/suites/0_stateless/02_00007_function_aggreator_on_empty.result:1]: Ok, In some databases, empty string is equal to null , like oracle.

**Comment by sundy-li** [comment]
@ zhaox1n  thanks for the contribution.

Some advice about this pr:

1. The state of the AggregateFunction should be initialized as Null, such as  `DataValueUInt64(None)` for count function.
2. To make the sql return null results on empty data, you just need to modify one line

```
--- a/fusequery/query/src/pipelines/transforms/transform_aggregator_final.rs
+++ b/fusequery/query/src/pipelines/transforms/transform_aggregator_final.rs
@@ -89,9 +89,9 @@ impl Processor for AggregatorFinalTransform {
         for func in &funcs {
             let merge_result = func.merge_result()?;
             // Check merge result null.
-            if merge_result.is_null() {
-                break;
-            }
+            // if merge_result.is_null() {
+            //     break;
+            // }
```

3.  Modify the ` pub fn to_array_with_size(&self, size: usize) -> Result<DataArrayRef> ` method to build array with  other None type values.  


4.  ` if input_rows == 0 ` is useless, we don't want this check in every impl functions.

**Comment by zhaox1n** [comment]
> @ zhaox1n thanks for the contribution.
> 
> Some advice about this pr:
> 
> 1. The state of the AggregateFunction should be initialized as Null, such as  `DataValueUInt64(None)` for count function.
> 2. To make the SQL return null results on empty data, you just need to modify 3 line codes.
> 
> ```
> --- a/fusequery/query/src/pipelines/transforms/transform_aggregator_final.rs
> +++ b/fusequery/query/src/pipelines/transforms/transform_aggregator_final.rs
> @@ -89,9 +89,9 @@ impl Processor for AggregatorFinalTransform {
>          for func in &funcs {
>              let merge_result = func.merge_result()?;
>              // Check merge result null.
> -            if merge_result.is_null() {
> -                break;
> -            }
> +            // if merge_result.is_null() {
> +            //     break;
> +            // }
> ```
> 
> 1. Modify the `pub fn to_array_with_size(&self, size: usize) -> Result<DataArrayRef>` method to build array with  other None type values, then we don't need transform_none_to_null anymore.
> 
> such as
> 
> ```
> -            DataValue::UInt64(Some(v)) => {
> -                Ok(Arc::new(UInt64Array::from(vec![*v; size])) as DataArrayRef)
> -            }
> +            DataValue::UInt64(v) => match v {
> +                Some(v) => Ok(Arc::new(UInt64Array::from(vec![*v; size])) as DataArrayRef),
> +                None => Ok(new_null_array(&DataType::UInt64, size)),
> +            },
> ```
> 
> 1. `if input_rows == 0` is useless, we don't want this check in every impl functions.
> 2. create our own `data_value_to_string` method to display the null value as MySQL behaviors.

CC @sundy-li @zhang2014

**Review by sundy-li — APPROVED** [review]
LGTM

**Comment by BohuTANG** [comment]
github api rejected the bot again

**Pull request merged**